<a href="https://colab.research.google.com/github/jPruim/cs344/blob/master/hw03/homeowrk3.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

From the class example. The entropy before the first decision is 1.

Price: 12 

  $: 7 
    y: 3
    n: 4
  entropy = -(3/7 ln(3/7) + 4/7 ln (4/7) = 0.6829081047
  $$: 2 

    y: 2
    n: 0
  entropy = 0

  $$$: 3 

    y: 1 
    n: 2
  entropy = -(1/3 * ln (1/3) + 2/3 * ln(2/3)) = 0.63651416829

avg: entropy = 7/12 * 0.683 + 2/12 * 0 + 3/12* 0.6365 = 0.55754166666

So info gain: 1- 0.558 = 0.442
Info gain from patrons is .541
However, Type gives no information about the situation. Thus, the Gain = 0. And therefore anything (including hungry) is atleast as good of a question.
However, Patrons is significantly better.
Thus Patrons > Price > Type





So in class we made a 2 layer network with 6 connections and 3 nodes

It is impossible to do in a single layer of linear neural net based on the fact that there is no "line" on the graph that separates the 1 and zeros of output. on a graph of inputs. 

If you were willing to except only values above zero and less than 1.
Then adding then 1 node with weights of 0.5 for both would do the trick.

Otherwise, it is impossible to simplify the neural net under the acceptance area of x>0. 





In [9]:
import keras
keras.__version__

from keras.datasets import boston_housing
(train_data, train_targets), (test_data, test_targets) =  boston_housing.load_data()

print( "Training data shape: " , train_data.shape)
print( "Test data shape: ", test_data.shape)


TensorFlow is already loaded. Please restart the runtime to change versions.
Training data shape:  (404, 13)
Test data shape:  (102, 13)


From lab example:
As you can see, we have 404 training samples and 102 test samples. The data comprises 13 features. The 13 features in the input data are as follow:

Per capita crime rate.
Proportion of residential land zoned for lots over 25,000 square feet.
Proportion of non-retail business acres per town.
Charles River dummy variable (= 1 if tract bounds river; 0 otherwise).
Nitric oxides concentration (parts per 10 million).
Average number of rooms per dwelling.
Proportion of owner-occupied units built prior to 1940.
Weighted distances to five Boston employment centres.
Index of accessibility to radial highways.
Full-value property-tax rate per $10,000.
Pupil-teacher ratio by town.
1000 * (Bk - 0.63) ** 2 where Bk is the proportion of Black people by town.
% lower status of the population.

In [32]:
#print(train_data[1])
#print(train_targets)
import math
import numpy as np

#Add new Charles River larger buildings variable: 1 if on river and rooms per dwelling above 15
#Add to training data
crl = []
for dataset in train_data:
  if dataset[3]*dataset[5]>6:
    crl.append([1])
  else:
    crl.append([0])  
print(crl)
train_data = np.append(train_data, crl, axis=1)
#print(train_data[1])
#print(len(train_data[1]))

#add variable to test data
crl = []
for dataset in test_data:
  if dataset[3]*dataset[5]>6:
    crl.append([1])
  else:
    crl.append([0])  
test_data = np.append(test_data, crl, axis=1)

validation_data = train_data[300:404]
validation_targets = train_targets[300:404]
train_data = train_data[0:299]
train_targets = train_targets[0:299]
#test data is the same as above




[[0], [0], [0], [0], [0], [0], [0], [0], [0], [0], [0], [0], [0], [0], [0], [0], [1], [0], [0], [0], [0], [0], [0], [0], [0], [0], [0], [0], [0], [0], [0], [0], [0], [0], [0], [0], [0], [0], [0], [0], [0], [0], [0], [0], [0], [0], [0], [0], [0], [0], [0], [0], [0], [0], [0], [0], [0], [0], [0], [0], [0], [0], [0], [0], [0], [0], [0], [0], [0], [0], [0], [0], [0], [0], [0], [0], [0], [0], [0], [0], [0], [0], [0], [0], [0], [0], [0], [0], [0], [0], [0], [0], [0], [0], [0], [0], [0], [0], [0], [0], [1], [0], [0], [0], [0], [0], [0], [0], [0], [0], [0], [0], [0], [0], [0], [0], [0], [0], [0], [0], [0], [0], [0], [0], [0], [0], [0], [0], [0], [0], [0], [0], [0], [0], [0], [0], [0], [0], [0], [0], [0], [0], [0], [0], [0], [0], [0], [0], [0], [0], [0], [0], [0], [0], [0], [0], [0], [0], [0], [0], [0], [0], [0], [1], [0], [0], [0], [0], [0], [0], [0], [0], [0], [0], [0], [0], [0], [0], [0], [0], [0], [0], [0], [0], [0], [0], [0], [0], [0], [1], [0], [0], [0], [0], [0], [0], [0], [0], [0], [0],

This variable might be useful (the crl variable) because it identifies zones that are on the river, and average a higher number of rooms per dwelling. This then could identify either apartment buildings or larger homesteads (depending on how rooms per dwelling is defined). Based on my interpretation, this could likely separate studio apratment zones from higher end apartments/residences.